lib/associations/belongs-to-many.js

```diff
@@ -415,41 +415,54 @@ module.exports = (function() {
     var association = this
       , primaryKeyAttribute = association.target.primaryKeyAttribute;
 
-    obj[this.accessors.set] = function(newAssociatedObjects, additionalAttributes) {
-      additionalAttributes = additionalAttributes || {};
-
-      if (newAssociatedObjects === null) {
-        newAssociatedObjects = [];
-      } else {
-        newAssociatedObjects = newAssociatedObjects.map(function(newAssociatedObject) {
-          if (!(newAssociatedObject instanceof association.target.Instance)) {
-            var tmpInstance = {};
-            tmpInstance[primaryKeyAttribute] = newAssociatedObject;
-            return association.target.build(tmpInstance, {
-              isNewRecord: false
-            });
-          }
-          return newAssociatedObject;
-        });
-      }
-
+    obj[this.accessors.set] = function(newAssociatedObjects, options) {
+      options = options || {};
       var instance = this;
 
       return instance[association.accessors.get]({}, {
-        transaction: additionalAttributes.transaction
+        transaction: options.transaction,
+        logging: options.logging
       }).then(function(oldAssociatedObjects) {
         var foreignIdentifier = association.foreignIdentifier
           , sourceKeys = Object.keys(association.source.primaryKeys)
           , targetKeys = Object.keys(association.target.primaryKeys)
           , obsoleteAssociations = []
           , changedAssociations = []
-          , defaultAttributes = additionalAttributes
-          , options = additionalAttributes
+          , defaultAttributes = options
           , promises = []
           , unassociatedObjects;
 
+        if (newAssociatedObjects === null) {
+          newAssociatedObjects = [];
+        } else {
+          if (!Array.isArray(newAssociatedObjects)) {
+            newAssociatedObjects = [newAssociatedObjects];
+          }
+          newAssociatedObjects = newAssociatedObjects.map(function(newAssociatedObject) {
+            if (!(newAssociatedObject instanceof association.target.Instance)) {
+              var tmpInstance = {};
+              tmpInstance[primaryKeyAttribute] = newAssociatedObject;
+              return association.target.build(tmpInstance, {
+                isNewRecord: false
+              });
+            }
+            return newAssociatedObject;
+          });
+        }
+
+        if (options.remove) {
+          oldAssociatedObjects = newAssociatedObjects;
+          newAssociatedObjects = [];
+          oldAssociatedObjects.filter(function(association) {
+            // Determine if this is an association we want to remove
+            return Utils._.find(oldAssociatedObjects, function(oldAssociatedObject) {
```

Comment:
Won't this always be true? You're matching an array against itself as far as i can tell?